File: frontend/types/universe.ts
/**
 * Universe TypeScript types — mirrors backend UniverseResponse schema.
 */

export type UniverseStatus = "draft" | "active" | "archived";

export type UniverseGenre =
  | "fantasy"
  | "science_fiction"
  | "mystery"
  | "horror"
  | "romance"
  | "adventure"
  | "historical"
  | "thriller"
  | "cyberpunk"
  | "steampunk"
  | "slice_of_life"
  | "other";

export interface Universe {
  id: string;
  name: string;
  slug: string;
  genre: UniverseGenre;
  description: string | null;
  tone: string | null;
  target_audience: string | null;
  status: UniverseStatus;
  cover_image: string | null;
  created_at: string;
  updated_at: string;
}

export interface UniverseCreatePayload {
  name: string;
  genre: UniverseGenre;
  description?: string;
  tone?: string;
  target_audience?: string;
}

export interface UniverseUpdatePayload {
  name?: string;
  genre?: UniverseGenre;
  description?: string;
  tone?: string;
  target_audience?: string;
  status?: UniverseStatus;
}

export interface UniverseListResponse {
  items: Universe[];
  total: number;
}

export const GENRE_LABELS: Record<UniverseGenre, string> = {
  fantasy: "Fantasy",
  science_fiction: "Science Fiction",
  mystery: "Mystery",
  horror: "Horror",
  romance: "Romance",
  adventure: "Adventure",
  historical: "Historical",
  thriller: "Thriller",
  cyberpunk: "Cyberpunk",
  steampunk: "Steampunk",
  slice_of_life: "Slice of Life",
  other: "Other",
};

File: backend/app/models/universe.py
from enum import StrEnum

from sqlalchemy import String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base
from app.database.mixins import BaseEntity


class UniverseStatus(StrEnum):
    DRAFT = "draft"
    ACTIVE = "active"
    ARCHIVED = "archived"
    draft = "draft"
    active = "active"
    archived = "archived"

class UniverseGenre(StrEnum):
    FANTASY = "fantasy"
    MYSTERY = "mystery"
    HORROR = "horror"
    THRILLER = "thriller"
    ADVENTURE = "adventure"
    ROMANCE = "romance"
    STEAMPUNK = "steampunk"
    CYBERPUNK = "cyberpunk"
    SCI_FI = "sci_fi"
    HISTORICAL = "historical"
    MODERN = "modern"
    OTHER = "other"
    fantasy = "fantasy"
    mystery = "mystery"
    horror = "horror"
    thriller = "thriller"
    adventure = "adventure"
    romance = "romance"
    steampunk = "steampunk"
    cyberpunk = "cyberpunk"
    sci_fi = "sci_fi"
    historical = "historical"
    modern = "modern"
    other = "other"

class Universe(BaseEntity, Base):
    __tablename__ = "universes"
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    slug: Mapped[str] = mapped_column(String(150), unique=True, nullable=False, index=True)
    genre: Mapped[str] = mapped_column(String(30), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    tone: Mapped[str | None] = mapped_column(String(200), nullable=True)
    target_audience: Mapped[str | None] = mapped_column(String(200), nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="draft")
    cover_image: Mapped[str | None] = mapped_column(String(500), nullable=True)
